UIKit/Location Controller/TWTLocationController.m

```diff
@@ -0,0 +1,189 @@
+//
+//  TWTLocationController.m
+//  Toast
+//
+//  Created by Steve Foster on 2/3/15.
+//  Copyright (c) 2015 Two Toasters, LLC. All rights reserved.
+//
+
+#import "TWTLocationController.h"
+
+@import CoreLocation;
+
+static NSTimeInterval const kLocationTimeoutInterval = 60;
+NSString *const kLocationControllerErrorDomain = @"locationcontroller";
+
+
+@interface TWTLocationController () <CLLocationManagerDelegate>
+
+@property (nonatomic, strong) CLLocationManager *locationManager;
+@property (nonatomic, strong) CLLocation *location;
+@property (nonatomic, strong) NSMutableArray *requestBlocks;
+
+@end
+
+@implementation TWTLocationController
+
++ (instancetype)sharedInstance
+{
+    static dispatch_once_t onceToken;
+    static TWTLocationController *sharedInstance = nil;
+    dispatch_once(&onceToken, ^{
+        sharedInstance = [[TWTLocationController alloc] init];
+    });
+    return sharedInstance;
+}
+
+
+- (instancetype)init
+{
+    self = [super init];
+    if (self) {
+        self.locationManager = [[CLLocationManager alloc] init];
+        self.locationManager.delegate = self;
+        self.locationManager.desiredAccuracy = kCLLocationAccuracyHundredMeters;
+    }
+    return self;
+}
+
+
+#pragma mark - Fetch Location
+
+- (void)fetchCurrentLocationWithCompletion:(TWTLocationControllerCompletion)completion
+{
+    self.locationCompletion = completion;
+
+    self.requestBlocks = [[NSMutableArray alloc] init];
+    [self.requestBlocks addObject:completion];
```

Comment:
Basically, what you want to do is determine if you have any requests running. The obvious way to do this is to check if request blocks is empty or not. Unfortunately, you need to do that in a thread-safe manner, so you’ll need to use `@synchronized` or a GCD queue or something. We can talk about it in the office.